src/SharpBrick.PoweredUp.BlueZ/BlueZPoweredUpBluetoothAdapter.cs

```diff
@@ -0,0 +1,142 @@
+
+            foreach (var propertyChanged in changes.Changed)
+            {
+                switch (propertyChanged.Key)
+                {
+                    case "Discovering":
+                        Discovering = (bool)propertyChanged.Value;
+                        break;
+                }
+            }
+        }
+
+        private async Task<ICollection<IDevice1>> GetExistingDevicesAsync()
+            => await Connection.System.FindProxies<IDevice1>();
+
+        private IDevice1 GetSpecificDeviceAsync(ObjectPath objectPath)
+            => Connection.System.CreateProxy<IDevice1>(BlueZConstants.BlueZDBusServiceName, objectPath);
+
+        private async Task<bool> IsLegoWirelessProcotolDevice(IDevice1 device)
+            => (await device.GetUUIDsAsync()).NullToEmpty().Any(x => x.ToUpperInvariant() == PoweredUpBluetoothConstants.LegoHubService);
+
+        public async void Discover(Func<PoweredUpBluetoothDeviceInfo, Task> discoveryHandler, CancellationToken cancellationToken = default)
+        {
+            _adapter ??= await GetAdapterAsync();
+
+            var existingDevices = await GetExistingDevicesAsync();
+
+            foreach (var device in existingDevices)
+            {
+                if (await IsLegoWirelessProcotolDevice(device))
+                {
+                    var poweredUpDevice = new BlueZPoweredUpBluetoothDevice(device, discoveryHandler);
+                    await poweredUpDevice.Initialize();
+
+                    _devices.Add(poweredUpDevice.DeviceInfo.BluetoothAddress, poweredUpDevice);
+
+                    await poweredUpDevice.TryGetManufacturerDataAsync();
+                }
+            }
+
+            await Connection.System.WatchInterfacesAdded(NewDeviceAddedHandler);
+
+            await _adapter.SetDiscoveryFilterAsync(new Dictionary<string,object>()
+            {
+                { "UUIDs", new string[] { PoweredUpBluetoothConstants.LegoHubService } }
+            });
+
+            cancellationToken.Register(async () =>
+            {
+                if (Discovering)
+                {
+                    await _adapter.StopDiscoveryAsync();
+                }
+            });
+
+            await _adapter.StartDiscoveryAsync();
+
+            async void NewDeviceAddedHandler((ObjectPath objectPath, IDictionary<string, IDictionary<string, object>> interfaces) args)
+            {
+                if (!args.interfaces.ContainsKey("org.bluez.Device1"))
+                {
+                    return;
+                }
+
+                var device = GetSpecificDeviceAsync(args.objectPath);
+                var poweredUpDevice = new BlueZPoweredUpBluetoothDevice(device, discoveryHandler);
+
+                await poweredUpDevice.Initialize();
+                
+                _devices.Add(poweredUpDevice.DeviceInfo.BluetoothAddress, poweredUpDevice);
+
+                await poweredUpDevice.TryGetManufacturerDataAsync();

```

Comment:
well, this is related to using certain property changes (RSSI in this case) as the trigger for the discovery handler callback. It could very well be that the manufacturer data comes along with other properties directly during the discovery. If the device is not added to the list before the discovery handler is invoked, the GetDeviceAsync will be called and it would not be able to return the device because it's not in the list of devices.

I know this is not pretty, plus come to think of it, adding it to the list might actually need to be done before the call to Initialize(), i'll have to check that to make sure it always works

Reply:
I had the same problem during my use of BlueGiga-adapter: The advertising-packets from the BLE-device(Hubs) are coming in an "unordered" sequence. You've got to fetch them all until you're sure to have alos the manufacturer-data (because this is needed in Bluetooth-kernel of poweredup).

Reply:
> I had the same problem during my use of BlueGiga-adapter: The advertising-packets from the BLE-device(Hubs) are coming in an "unordered" sequence. You've got to fetch them all until you're sure to have alos the manufacturer-data (because this is needed in Bluetooth-kernel of poweredup).

I'm not familiar with the bluegiga adapter, does that even use bluez? i'm unsure if unordered sequence of advertisement packets is an actual issue here

Reply:
No, the BlueGiga-adapter ("dongle") BLED112 is a serial-device attached to USB. The client (poweredup-program here) talks to it just over a Serial-object. So it doesn't need Bluez or any other BLE-stack in the operating system, because the device itself handles the BLE-communication (timing and alike)